rbd/watchers_mimic.go

```diff
@@ -59,3 +76,77 @@ func (image *Image) ListWatchers() ([]ImageWatcher, error) {
 	}
 	return imageWatchers, nil
 }
+
+// watchCallbacks tracks the active callbacks for rbd watches
+var watchCallbacks = callbacks.New()
+
+// WatchCallback defines the function signature needed for the UpdateWatch
+// callback.
+type WatchCallback func(interface{})
+
+type watchInstance struct {
```

Comment:
On the naming: Yeah, I wasn't happy with that naming when I added it either. How about "watchItem"? It is split out so that the "callback tracking" is only done in the 'Watch' type and the callback registry only has this "inner" type for values (note that the Watch types needs the int registry key).

On data arguments: it could have been done that way but this approach more closely matches what was previously done for the diff_iterate callback support and how ceph does it. There's also a (weak) argument to be had that creating lots of callback functions might be less efficient than one function with data arguments, but I didn't measure anything so it's conjecture on my part. If it's not a big deal I think we ought to keep it for consistency with the ceph apis.

Reply:
Naming: How about `callbackItem`?

Passthru args: First of all: sorry, I wrote "unary", but of course meant "nullary", because it is unary already.
So you think creating an `watchInstance{ callback: cb, data: data }` is more expensive than creating a `func() { cb(data) }`? Would surprise me, but I also don't know. The big "pros" would be for me: the `data` doesn't have to leave go space, we would have type safety and no weird casting is required. You could simply use `type WatchCallback func()` then.

Reply:
Naming: How about `callbackItem`?

Pass-thru args: First of all: sorry, I wrote "unary", but of course meant "nullary", because it is unary already.
So you think creating an `watchInstance{ callback: cb, data: data }` is more expensive than creating a `func() { cb(data) }`? Would surprise me, but I also don't know. You could simply use `type WatchCallback func()` then.

When I first wrote this answer I also wrote 
> The big "pros" would be for me: the `data` doesn't have to leave go space, we would have type safety and no weird casting is required.

but then realized that I was mixing up the _two_ pass-thru arguments here: the `data` of the go-callBack, and the `arg` of the C-callback, that you are using for the `cbIndex`. So my first thought was to actually not use _that_ argument, but now I'm not sure anymore if that is possible, because I'm not sure we can expose anonymous go functions to C-callback-apis. 

So my question in the end would be: can we directly pass an anonymous nullary go-callback to the C-api, and simply just store the returned handle together with the `image` in a registry? But looking at your comments in `callbacks.go` this is exactly not possible, and we _need_ the arg for the index. I realize that I probably go through the very same thought process as you did before. Sorry about that. :man_facepalming: